build/scripts/package.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-nocheck <-- This is needed as this file is basically a JavaScript script
+//                 but with some TypeScript niceness baked in
+const fs = require('fs')
+const path = require('path')
+const glob = require('glob')
+const execSync = require('child_process').execSync
+const pkg_dir = 'dist/package'
+
+async function copyGlob(pattern) {
+  glob(pattern, (error, files) => {
+    for (var i = 0; i < files.length; i++) {
+      // Skip dirs as they are most likely listed in the .vscodeignore as well
+      if (fs.statSync(files[i]).isDirectory()) continue
+
+      let src = files[i]
+      let dst = path.join(pkg_dir, src.replace('build/package', ''))
```

Comment:
I'd prefer we don't hardcode `build/package` in this function. I imagine we might want to use this for something else in the future. Did the cwd approach mentioned in another comment not work?

Reply:
> Do you want maybe an entry to add to package.json like packageDir and this value is used to get the pkg_dir?

It's more that I dont' think this function show know about the build/package directory at all.

For example, what if we also wanted

```
copyGlob("build/other-stuff/*")
```

Now we need to special case build/other-stuff too. Just trying to make this as generic as possible so we don't ever run into edge cases.

Reply:
> Would something like be better, having a optional param `copyToRoot` and this is how the file is determined
> 
> ```ts
> let dst = copyToRoot
>         ? path.join(pkg_dir, src.replace(path.dirname(src), ''))
>         : path.join(pkg_dir, src)
> ```

That removes the entire directory part of the src though, leaving just the filename, right? We might want directories as well. I previously comment an alternative copyGlob idea that takes a dir parameter to tell what the root is, thoughts on something like this:

```
function copyGlob(pattern, dir = ".") {
  glob(pattern, {cwd=dir}, (error, files) => {
    for (var i = 0; i < files.length; i++) {
      let src = path.join(dir, file[i])
      let dst = path.join(pkg_dir, file[i])
      let parentDir = path.dirname(dst)
      ...  // ensure parentDir exists
      fileCopy(src, dst)
    }
  })
}
```

Then the usage is 
```
copyGlob("*", "build/package")
```
But also allows more complex things like

```
copyGlob("foo/*", "build/bar")
```
This will copy everything in `build/bar` that matches `foo/*`, including the `foo` directory. I think the copyToRoot idea doesn't do that and would just copy everything in foo/* without the foo directory.

It's a similar idea, but gives more control over what is the root and what isn't.

Reply:
So this makes sense but for the files inside of `build/package` don't we want them directly inside of `dist/package` and not `dist/package/build/package`?

Reply:
If we do something like `copyGlob("*", "build/package")`, then `cwd` is set to `build/`  and the glob is run from there. That way the list of `files` won't include `build/package/`. That's why we then need to add `dir` to get the source file and `pkg_dir` to get the dst file.

For example, if we had `copyGlob("foo/*", "build/package")`, then `files` would be something like `[foo/bar.txt, foo/baz.txt]`. We then want to copy "dir + / + those" (e.g. build/package/foo/bar.txt) to "pkg_dir + / + those" (e.g. dist/package/foo/bar.txt). This way we keep the foo/ part of the glob but exclude build/package.